src/storage/v2/vertex_accessor.cpp

```diff
@@ -616,16 +618,38 @@ Result<EdgesVertexAccessorResult> VertexAccessor::InEdges(View view, const std::
   {
     auto guard = std::shared_lock{vertex_->lock};
     deleted = vertex_->deleted;
-    expanded_count = static_cast<int64_t>(vertex_->in_edges.size());
-    // TODO: a better filter copy
-    if (edge_types.empty() && !destination) {
-      in_edges = vertex_->in_edges;
+    if (hops_limit && hops_limit->has_value()) {
+      if (edge_types.empty() && !destination) {
+        if (**hops_limit <= 0 && static_cast<int64_t>(vertex_->in_edges.size()) > 0) {
+          *hops_limit = -1;
+        } else {
+          expanded_count = (std::min(**hops_limit, static_cast<int64_t>(vertex_->in_edges.size())));
+          **hops_limit -= expanded_count;
+          std::copy_n(vertex_->in_edges.begin(), expanded_count, std::back_inserter(in_edges));
+        }
+      } else {
+        for (const auto &[edge_type, to_vertex, edge] : vertex_->in_edges) {
+          --(**hops_limit);
+          if (**hops_limit < 0) break;
+          expanded_count++;
+          if (destination && to_vertex != destination_vertex) continue;
+          if (!edge_types.empty() && std::find(edge_types.begin(), edge_types.end(), edge_type) == edge_types.end())
+            continue;
+          in_edges.emplace_back(edge_type, to_vertex, edge);
+        }
+      }
     } else {
-      for (const auto &[edge_type, from_vertex, edge] : vertex_->in_edges) {
-        if (destination && from_vertex != destination_vertex) continue;
-        if (!edge_types.empty() && std::find(edge_types.begin(), edge_types.end(), edge_type) == edge_types.end())
-          continue;
-        in_edges.emplace_back(edge_type, from_vertex, edge);
+      expanded_count = static_cast<int64_t>(vertex_->in_edges.size());
+      // TODO: a better filter copy
```

Comment:
what is a better filter copy?

Reply:
That comment isn't placed here by me. Was here before but since I changed this file it looks like I have put it.
I also don't get it, maybe we can delete it.